rust/foxglove-derive/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, Data, DeriveInput, Fields};
+
+/// Derive macro for enums and structs allowing them to be logged to a Foxglove channel.
+#[proc_macro_derive(Loggable)]
+pub fn derive_loggable(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    match &input.data {
+        Data::Enum(_) => derive_enum_impl(input),
+        Data::Struct(_) => derive_struct_impl(input),
+        _ => panic!("Loggable can only be used with enums or structs"),
+    }
+}
+
+fn derive_enum_impl(input: DeriveInput) -> TokenStream {
+    let name = &input.ident;
+
+    // Extract variants from the enum
+    let variants = match &input.data {
+        Data::Enum(data) => &data.variants,
+        _ => unreachable!(),
+    };
+
+    // Generate variant name and number pairs for enum descriptor
+    let variant_descriptors = variants.iter().enumerate().map(|(i, v)| {
+        let variant_name = v.ident.to_string();
+        let variant_value = i as i32;
+        quote! {
+            let mut value = foxglove::prost_types::EnumValueDescriptorProto::default();
+            value.name = Some(String::from(#variant_name));
+            value.number = Some(#variant_value);
+            enum_desc.value.push(value);
+        }
+    });
+
+    // Generate implementation
+    let expanded = quote! {
+        impl foxglove::ProtobufField for #name {
+            fn field_type() -> foxglove::prost_types::field_descriptor_proto::Type {
+                foxglove::prost_types::field_descriptor_proto::Type::Enum
+            }
+
+            fn wire_type() -> u32 {
+                0 // Varint, same as integers
+            }
+
+            fn write(&self, buf: &mut impl foxglove::bytes::BufMut) {
+                let mut value = *self as u64;
+                while value >= 0x80 {
+                    buf.put_u8((value as u8) | 0x80);
+                    value >>= 7;
+                }
+                buf.put_u8(value as u8);
```

Comment:
This should be i32 to match the descriptor type. Maybe also reuse the existing implementation:

```suggestion
                let value = *self as i32;
                ::foxglove::ProtobufField::write(&value, buf);
```

Reply:
I don't understand the i32 comment; this should be writing a varint, right? In #396 I re-used the varint fn as suggested above.